fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/ComparisonRange.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * ComparisonRange.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.temp;
+
+import com.apple.foundationdb.API;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.query.expressions.Comparisons;
+import com.apple.foundationdb.record.query.plan.ScanComparisons;
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * A set of compatible comparisons on a single field of a{@link com.apple.foundationdb.record.metadata.expressions.KeyExpression}
+ * representing a contiguous range of values for that field.
+ *
+ * <p>
+ * A {@code ComparisonRange} is similar to a {@link ScanComparisons} but represents a contiguous range of values for a
+ * single field, rather than for an entire {@code KeyExpression}. A comparison range is effectively a sum type, with
+ * the following options:
+ * </p>
+ * <ul>
+ *     <li>An empty set of comparisons, indicating that the entire universe of possible values is in the range.</li>
+ *     <li>
+ *         An equality comparison on the field. There can only be a single such comparison. For example, single
+ *         comparison of the form {@code EQUALS "foo"}.
+ *     </li>
+ *     <li>
+ *         A set of inequality comparisons that define a single contiguous range of values. For example, the combination
+ *         of the comparisons {@code > 8} and {@code < 30}. The comparison range may include redundant
+ *         comparisons by contract, but it may or may not simplify the range into a more compact form. For example,
+ *         the comparison range can include the comparisons {@code > 8}, {@code < 30}, and {@code EQUALS 10}, but it
+ *         may optionally simplify this to {@code > 8} and {@code < 30}. Note that this behavior is not fully
+ *         implemented right now; the current implementation might accept a set of comparisons that do not define a
+ *         contiguous range such as {@code < 8} and {@code > 30}. However, this validation logic will be added here in
```

Comment:
If this were handled, would it mean two ranges or empty?

Reply:
My mental model here is that a `ComparisonRange` is something that can be turned into a single `RecordQueryIndexPlan` (or `RecordQueryScanPlan`, I suppose). In this case, you'd need a union or something of two plans (I think), so it would be two ranges.

Reply:
With some more careful thought, I think it should just be empty, and handling this should actually go in this class. I'll clarify that the range normalization is the unimplemented part.

Reply:
Okay, so this class represents a single, contiguous range, perhaps derived from complex, overlapping comparisons.

Something else normalizes all comparisons into a (minimal?) set of comparison ranges. Is that the (eventual) idea?

Reply:
Yes, that's the idea of this class. In particular, that's how it's used in `KeyExpressionComparisons`, which is currently the only place that uses it. Eventually I could imagine moving a bunch of logic that's currently in `ScanComparisons` into here so that it only needs to be implemented once.

I think it depends on what you mean by "something else". I think that the interface of `ComparisonRange` should always boil down to "here is a single, contiguous range; can I add this comparison to it?". So the normalization logic probably lives here, but the logic for "deciding" which parts of a comparison should go into each of several possible ranges (as for intersection/union planning) should go elsewhere, I think.

Reply:
Does that mean splitting `NOT_EQUALS x` into `LESS_THAN x OR GREATER_THAN x` somewhere in the future?

Reply:
Probably, in some circumstances (I don't know when and where, though, just that it would be handled by a planner rule rather than `ComparisonRange`).